Add implementations for cutset networks (CNets)

yangyang-tue opened the pull request:
Adding implementations for cutset networks (CNets), including:

- A BinaryCNet class, which implements the classical CNet learning algorithm based on the information gain heuristic from decision tree literature.
- The learning algorithm based on the Bayesian-Dirichlet equivalent uniform (BDeu) score.
- The learning algorithm based on the Bayesian information criterion (BIC) score.
- Functions to compute counts of RVs given some data sets.
- Functions to estimate Bayesian posterior parameters.

loreloc commented:
Hi, thank you for your PR!

I have a couple of points to make about these commits.
- Some introduced methods lack of proper documentation. Can you add that?
- Is there any reason why you introduced ```estimate_priors_joints_bayesian()``` instead of using the existing ```estimate_priors_joints()``` in ```deeprob.utils.statistics```?

yangyang-tue commented:
Hi @loreloc ,

Thank you for your reply!

- I have added the missing documentation for all the methods :)
- Actually, I introduce the function `estimate_priors_joints_bayesian()` for two reasons: 1) to avoid confusion between the Laplace smoothing and Bayesian posterior computation; 2) to introduce an additional parameter `fracs` to support the structural EM algorithm. 
But now I think it's indeed better to remove that to avoid duplicate code. Later I will try to integrate this feature as well as the implementation of the structural EM algorithm in other commits.
- I also added a small example for learning a CNet using the BD score.
